### PR TITLE
[sw, dif_uart] Add API for UART RX timeout control

### PR DESCRIPTION
**Pull request by mundaym**
Add the function dif_uart_set_timeout to allow the RX timeout to be
enabled and disabled as required.

Updates #4277.

### CODE REVIEW

File: sw/device/lib/dif/dif_uart.c [hunk 1]
```diff
@@ -539,3 +539,25 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
 
   return kDifUartOk;
 }
+
+dif_uart_result_t dif_uart_set_timeout(const dif_uart_t *uart, uint32_t value,
+                                       dif_uart_toggle_t enable) {
+  if (uart == NULL || (value & ~UART_TIMEOUT_CTRL_VAL_MASK) != 0) {
```
Comment (mundaym) [inline sw/device/lib/dif/dif_uart.c:545]: Not sure if this is the correct way to handle out of range timeout values.

**Comment by mundaym** [comment]
A few questions I have about this (apologies if they are obvious, this is my first DIF change):

1. Would it be preferable to split the `set_timeout` API into separate `enable` and `disable` functions?
2. Is it correct to return a 'bad arg' error if the timeout is out of range or should that be handled differently?
3. Do we need a way to read the timeout value? Currently the API only allows it to be set.
4. Does this require a smoke test (or other hardware test) or are the unit tests sufficient?

**Comment by lenary** [comment]
> A few questions I have about this (apologies if they are obvious, this is my first DIF change):
> 
> 1. Would it be preferable to split the `set_timeout` API into separate `enable` and `disable` functions?

Yes, I think this makes things clearer, and shows that you don't need to choose a timeout length when you're disabling the timeout.

> 2. Is it correct to return a 'bad arg' error if the timeout is out of range or should that be handled differently?

Bad Arg is fine, if you want to return a more specific error, that's ok too!

> 3. Do we need a way to read the timeout control register value? Currently the API only allows it to be set.

Yes, there should be a read API, as no peripheral accesses are allowed that don't use the DIF.

> 4. Does this require a smoke test (or other hardware test) or are the unit tests sufficient?

Unit tests are sufficient.

**Review by silvestrst — COMMENTED** [review]
@mundaym thanks!

Looking good!

I have couple minor comments, and if you could implement @lenary suggested changes - I think it then can get merged!

Separate thanks for your help in pushing this change to close of "DIF_HW_USAGE_REVIEWED" item.

File: sw/device/lib/dif/dif_uart.c (hunk 1 above)
Comment (mcy) [inline sw/device/lib/dif/dif_uart.c:545]: There should be a test for the mask check here.

File: sw/device/lib/dif/dif_uart.c [hunk 2]
```diff
@@ -539,3 +539,25 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
 
   return kDifUartOk;
 }
+
+dif_uart_result_t dif_uart_set_timeout(const dif_uart_t *uart, uint32_t value,
+                                       dif_uart_toggle_t enable) {
```
Comment (mcy) [inline sw/device/lib/dif/dif_uart.c:544]: Are you required the pass a value to disable the timeout? Maybe it makes more sense to have separate `enable_timeout` and `disable_timeout` functions, where the former accepts a timeout value.

Alternatively, require that for disable, a specific `value` is used?

File: sw/device/lib/dif/dif_uart.h [hunk 3]
```diff
@@ -532,6 +532,18 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
...
+/**
+ * Enables or disables the RX timeout.
+ *
+ * @param uart A UART handle.
+ * @param value RX timeout value in UART bit times in the range [0,0xffffff].
+ * @param enable Enable RX timeout feature.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_uart_result_t dif_uart_set_timeout(const dif_uart_t *uart, uint32_t value,
```
Comment (mcy) [inline sw/device/lib/dif/dif_uart.h:544]: Maybe `_set_rx_timeout`?

File: sw/device/lib/dif/dif_uart.c (hunk 1 above)
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.c:545]: There is a unit test that checks that out-of-range values are rejected. I could add more testing of in-range values, though the EnableDisable test does already check that the max valid value is accepted.

File: sw/device/lib/dif/dif_uart.h [hunk 4]
```diff
@@ -532,6 +532,18 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
                                         dif_uart_loopback_t loopback,
                                         dif_uart_toggle_t enable);
 
+/**
+ * Enables or disables the RX timeout.
+ *
+ * @param uart A UART handle.
+ * @param value RX timeout value in UART bit times in the range [0,0xffffff].
```
[the remark this answers is not in the document]
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.h:539]: `timeout` or perhaps `duration` seem like reasonable names. How closely should I follow the names used in the documentation for the control register (that's where `value` comes from)?

[the remark this answers is not in the document]
Comment (lenary) [reply sw/device/lib/dif/dif_uart.h:539]: You don't need to follow the field names in the control register if you think they're unclear. One thing that would be an improvement is suffixing with the unit (so `timeout_ticks` or `duration_ticks`).

File: sw/device/lib/dif/dif_uart.h (hunk 3 above)
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.h]: Done (as `_{enable,disable,get}_rx_timeout`).

File: sw/device/lib/dif/dif_uart.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.h]: Renamed to `duration_ticks`.

File: sw/device/lib/dif/dif_uart.c [hunk 5]
```diff
@@ -539,3 +539,25 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
...
+  if (uart == NULL || (value & ~UART_TIMEOUT_CTRL_VAL_MASK) != 0) {
+    return kDifUartBadArg;
+  }
+
+  uint32_t reg = 0;
+  if (enable == kDifUartToggleEnabled) {
+    reg = bitfield_field32_write(reg,
```
[the remark this answers is not in the document]
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.c]: Done. Should we also (in a different PR) do that for the other functions in this file?

File: sw/device/lib/dif/dif_uart.c (hunk 2 above)
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.c]: Added separate `enable` and `disable` functions. The `disable` function as written will reset the timeout value (duration) to 0 ticks. This isn't strictly necessary but seems reasonable. If needed the preexisting timeout value can be saved using `_get_rx_timeout` before disabling the timeout.

File: sw/device/lib/dif/dif_uart.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (silvestrst) [reply sw/device/lib/dif/dif_uart.c]: Don't worry about changing other places in the library that are not relevant to this change. It is a boring and low priority clean-up that I imagine will happen on DIF to S2 transition (or around that time).

**Comment by mundaym** [comment]
Thanks for the reviews! A few questions:

1. I don't love the `_get_rx_timeout` name but can't immediately think of anything better. Any suggestions or is it best to leave as is?
2. I've called the EN bit out paramter `status` in the `get` function. Would it be clearer or more idiomatic to call it something else (`enable`, `enabled`, or `toggle` maybe)?
3. In the `get` function I've made the `status` out parameter mandatory and the `duration_ticks` out parameter optional. My thinking is that there might be situations where you want to check if the timeout is enabled but don't care about the duration. Does this seem reasonable? I could just make them both mandatory.

File: sw/device/lib/dif/dif_uart.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (imphil) [reply sw/device/lib/dif/dif_uart.c]: (Or just do it when you have a couple of minutes of free time.)

**Review by lenary — APPROVED** [review]
LGTM! Thanks @mundaym!

**Comment by lenary** [comment]
> Thanks for the reviews! A few questions:
> 
> 1. I don't love the `_get_rx_timeout` name but can't immediately think of anything better. Any suggestions or is it best to leave as is?

I'm happy with it as-is 😄 

> 2. I've called the EN bit out paramter `status` in the `get` function. Would it be clearer or more idiomatic to call it something else (`enable`, `enabled`, or `toggle` maybe)?

In the DIF template, we've called this `state`, but I'm happy with `status` too.

> 3. In the `get` function I've made the `status` out parameter mandatory and the `duration_ticks` out parameter optional. My thinking is that there might be situations where you want to check if the timeout is enabled but don't care about the duration. Does this seem reasonable? I could just make them both mandatory.

I think this is reasonable. Sometimes we've documented when out parameters will or won't be updated, but I'm happy that the docs match the implementation (which always writes to the out parameter if it is present).

File: sw/device/lib/dif/dif_uart.c [hunk 6]
```diff
@@ -539,3 +539,50 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
...
+  return kDifUartOk;
+}
+
+dif_uart_result_t dif_uart_disable_rx_timeout(const dif_uart_t *uart) {
+  if (uart == NULL) {
+    return kDifUartBadArg;
+  }
+
+  uint32_t reg = bitfield_bit32_write(0, UART_TIMEOUT_CTRL_EN_BIT, false);
+  reg = bitfield_field32_write(reg, UART_TIMEOUT_CTRL_VAL_FIELD, 0);
```
Comment (silvestrst) [inline sw/device/lib/dif/dif_uart.c:564]: I think you probably don't need this step as it can be enabled only via `*enable*` function, and hence will set the `VAL` field to appropriate value.

File: sw/device/lib/dif/dif_uart.c [hunk 7]
```diff
@@ -539,3 +539,50 @@ dif_uart_result_t dif_uart_loopback_set(const dif_uart_t *uart,
...
+                                          uint32_t *duration_ticks) {
+  if (uart == NULL || status == NULL) {
+    return kDifUartBadArg;
+  }
+
+  uint32_t reg =
+      mmio_region_read32(uart->params.base_addr, UART_TIMEOUT_CTRL_REG_OFFSET);
+  *status = bitfield_bit32_read(reg, UART_TIMEOUT_CTRL_EN_BIT)
+                ? kDifUartToggleEnabled
+                : kDifUartToggleDisabled;
```
Comment (silvestrst) [inline sw/device/lib/dif/dif_uart.c:582]: Nit: could you please insert a newline between ternary and `if`? Seem like two unrelated code blocks.

**Review by silvestrst — APPROVED** [review]
@mundaym thanks!

Just couple minor comments bellow, otherwise LGTM!

File: sw/device/lib/dif/dif_uart.c (hunk 6 above)
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.c:564]: I included this as documentation to make it explicit that the intention is to 0 the VAL rather than leave it unchanged. We can get rid of it if you think that is unnecessary. We could also just set the value of the register to 0 (the reset value) which is what we're really doing anyway.

File: sw/device/lib/dif/dif_uart.c (hunk 7 above)
Comment (mundaym) [reply sw/device/lib/dif/dif_uart.c:582]: Done
